src/socket.rs

```diff
@@ -39,20 +39,8 @@ pub fn set_transparent(l: &TcpListener) -> io::Result<()> {
     Ok(())
 }
 
-#[cfg(target_os = "linux")]
-#[allow(unsafe_code)]
-pub fn orig_dst_addr_fd<T: AsRawFd>(sock: T) -> io::Result<SocketAddr> {
-    let fd = sock.as_raw_fd();
-
-    unsafe { linux::so_original_dst(fd) }
-}
-
-#[cfg(not(target_os = "linux"))]
-pub fn orig_dst_addr_fd<T: AsRawFd>(_: T) -> io::Result<SocketAddr> {
-    Err(io::Error::new(
-        io::ErrorKind::Other,
-        "SO_ORIGINAL_DST not supported on this operating system",
-    ))
+pub fn orig_dst_addr_or_default(sock: &tokio::net::TcpStream) -> SocketAddr {
+    orig_dst_addr(sock).unwrap_or_else(|_| sock.peer_addr().expect("must get peer address"))
```

Comment:
I've been looking at this code, and I cannot wrap my head around the logic here, but perhaps I'm missing something.

In case of the lookup failure, this will default to a `peer_addr`, which is the IP of the connecting client.

This means that later we proxy traffic between `remote_addr` source (which is `peer_addr`) and `orig` destination (which is also `peer_addr`), which cannot be right.

@howardjohn could you explain?

Reply:
I think this is probably inverted and it isn't breaking things (yet) since our tests either run to/from localhost OR run with actual iptables (in which case fallback never occurs)

Reply:
One thing to note is that even if there is no redirect, the orig_dst lookup succeeds - it will just return the same address (local_peer); the only case it fails is if there are 0 iptables rules on the node entirely

Reply:
Right, but we could compare `orig_dst` and `local_addr` and bail out in case they're the same.

But we can do that only for `REDIRECT`s and/or when running `IP_TRANSPARENT` fails, since `TPROXY` overrides `local_addr` with `orig_dst`, and I don't think we can detect when that have happened.

Reply:
PR with the `peer_addr` -> `local_addr` change: #234.